app/services/solidus_bolt/users/add_payment_source_service.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module SolidusBolt
+  module Users
+    class AddPaymentSourceService < SolidusBolt::BaseService
+      attr_reader :user, :payment_methods
+
+      def initialize(user:, payment_methods:)
+        @user = user
+        @payment_methods = payment_methods
+        super
+      end
+
+      def call
+        payment_methods.each do |payment_method|
+          add_payment_source(payment_method)
+        end
+      end
+
+      private
+
+      def add_payment_source(payment_method)
+        # format of card_expiration: '2022-04'
+        card_expiration = "#{payment_method['exp_year']}-#{payment_method['exp_month'].to_s.rjust(2, '0')}"
+        payment_source = SolidusBolt::PaymentSource.find_or_create_by!(
+          user_id: user.id,
+          card_last4: payment_method['last4'],
+          card_expiration: card_expiration
+        )
+        return if payment_source.card_id
+
+        payment_source.update(card_id: payment_method['id'])
```

Comment:
Why do you have to update the card_id in another place?
I think that user_id, last, expiration and id should be the unique

```
payment_source = SolidusBolt::PaymentSource.find_or_create_by!(
          user_id: user.id,
          card_last4: payment_method['last4'],
          card_expiration: card_expiration,
          card_id: payment_method['id']
        )
```

Reply:
after discussing internally we agreed that this service should find or create in all cases

commit 57d29313dff9bbab7d779c4fe648e1aacfa41b3e is implemented with this in mind